### PR TITLE
Migrate to XML Based Configuration

### PR DESCRIPTION
**Pull request by os-d**
# Preface

Please ensure you have read the [contribution docs](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md) prior
to submitting the pull request. In particular,
[pull request guidelines](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md#pull-request-best-practices).

## Description

mu_feature_config has moved to an XML based configuration system. This PR integrates changes from that repo and mu_oem_sample to demonstrate and run tests on an end to end XML based configuration system.

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [x] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

Tested manually by changing config knobs in XML and seeing different behavior, dumping config, applying config via SVD in ConfApp.

## Integration Instructions

N/A, this is the example integration for a platform.

### CODE REVIEW

File: Features/mu_config_ext_dep.yaml [hunk 1]
```diff
@@ -1,14 +1,12 @@
...
+# This external dependency pulls mu_feature_config for platforms to consume.
 #
 {
   "scope": "global",
   "type": "git",
   "name": "FEATURE_CONFIG",
   "var_name": "FEATURE_CONFIG_PATH",
-  "source": "https://github.com/microsoft/mu_feature_config.git",
-  "version": "aedb83bef7882353b3ee36958f23df65468db3eb",
+  "source": "https://github.com/kuqin12/mu_feature_config.git",
```
Comment (kuqin12) [inline Features/mu_config_ext_dep.yaml:9]: Just keep it as a stub that we need to change this once we finalize the config repo update.

File: Features/mu_config_ext_dep.yaml [hunk 2]
```diff
@@ -1,14 +1,12 @@
...
+  "version": "54d79dd425c201000457b4e648448d936706068f",
   "flags": ["set_build_var"]
-}
+}
```
Comment (kuqin12) [inline Features/mu_config_ext_dep.yaml:12]: nit: somehow no new line?

File: Platforms/QemuQ35Pkg/ConfigKnobs/ConfigKnobs.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  Publish default silicon policy for GFX to policy database.
+
+  @param FileHandle                     The image handle.
+  @param PeiServices                    The PEI services table.
+
+  @retval Status                        From internal routine or boot object, should not fail
+**/
+EFI_STATUS
+EFIAPI
+PublishDefaultPolicyGfx (
```
Comment (kuqin12) [inline Platforms/QemuQ35Pkg/ConfigKnobs/ConfigKnobs.h:22]: Remove. This is in a separate module now.

File: Platforms/QemuQ35Pkg/Library/ConfigSystemModeLibQ35/ConfigSystemModeLib.c [hunk 4]
```diff
@@ -25,5 +25,5 @@ IsSystemInManufacturingMode (
   VOID
   )
 {
-  return (GetMfciSystemOperationMode () == OEM_UEFI_MANUFACTURING_MODE);
+  return /*(GetMfciSystemOperationMode () == OEM_UEFI_MANUFACTURING_MODE)*/ TRUE;
```
Comment (kuqin12) [inline Platforms/QemuQ35Pkg/Library/ConfigSystemModeLibQ35/ConfigSystemModeLib.c:28]: Should not be here :)

File: Platforms/QemuQ35Pkg/Library/Q35ConfigDataLib/Q35ConfigDataLib.inf [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+##
+
+
+[Defines]
+  INF_VERSION         = 0x00010017
+  BASE_NAME           = Q35ConfigDataLib
+  FILE_GUID           = 25D1A00F-E85F-4F1D-8251-02F3E4EDAB1C
+  VERSION_STRING      = 1.0
+  MODULE_TYPE         = BASE
+  LIBRARY_CLASS       = PlatformConfigDataLib
```
Comment (kuqin12) [inline Platforms/QemuQ35Pkg/Library/Q35ConfigDataLib/Q35ConfigDataLib.inf:16]:
```suggestion
  LIBRARY_CLASS       = NULL
```
I think this is more of a "NULL" library class?

File: Platforms/QemuQ35Pkg/Plugins/QemuRunner/QemuRunner.py [hunk 6]
```diff
@@ -109,7 +109,7 @@ def Runner(env):
         gdb_port = env.GetValue("GDB_SERVER")
         if (gdb_port != None):
             logging.log(logging.INFO, "Enabling GDB server at port tcp::" + gdb_port + ".")
-            args += " -gdb tcp::" + gdb_port
+            args += " -S -gdb tcp::" + gdb_port
```
Comment (kuqin12) [inline Platforms/QemuQ35Pkg/Plugins/QemuRunner/QemuRunner.py:112]: nit: do we want this to be added to mainline?

File: Platforms/QemuQ35Pkg/ConfigKnobs/ConfigKnobs.inf [hunk 7]
```diff
@@ -8,39 +8,41 @@
...
   VERSION_STRING      = 1.0
   MODULE_TYPE         = PEIM
-  ENTRY_POINT         = ConfigDataGfxEntry
+  ENTRY_POINT         = ConfigKnobsEntry
 
 [Sources]
-  ConfigDataGfx.c
+  ConfigKnobs.c
+  ConfigKnobs.h
+  PolicyDataGfx/PolicyDataGfx.c
```
Comment (kuqin12) [inline Platforms/QemuQ35Pkg/ConfigKnobs/ConfigKnobs.inf:20]: This line and the corresponding file should be gone?

File: Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc [hunk 8]
```diff
@@ -1184,12 +1175,8 @@ PlatformSmmProtectionsTestLib|UefiTestingPkg/Library/PlatformSmmProtectionsTestL
   # of FrontPage to be developed and tested while RngLib or other parts of the authentication process are being developed.
   DfciPkg/IdentityAndAuthManager/IdentityAndAuthManagerDxe.inf
 
-  # Processes ingoing and outgoing DFCI settings requests.
   DfciPkg/DfciManager/DfciManager.inf
```
Comment (kuqin12) [inline Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc:1178]: nit: why was this comment removed?

File: Platforms/QemuQ35Pkg/Library/Q35ConfigDataLib/Q35ConfigDataLib.inf (hunk 5 above)
Comment (os-d) [reply Platforms/QemuQ35Pkg/Library/Q35ConfigDataLib/Q35ConfigDataLib.inf]: Yep, good point.

File: Platforms/QemuQ35Pkg/Plugins/QemuRunner/QemuRunner.py (hunk 6 above)
Comment (os-d) [reply Platforms/QemuQ35Pkg/Plugins/QemuRunner/QemuRunner.py]: Maybe, but not on this PR.

File: Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc (hunk 8 above)
Comment (os-d) [reply Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc:1182]: No clue, put it back.

File: Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  IN      POLICY_PROTOCOL  *PolicyInterface,
+  IN      BOOLEAN          PlatformGfxPowerOnPort0,
+  OUT     GFX_POLICY_DATA  *GfxSiliconPolicy OPTIONAL,
+  IN OUT  UINT16           *PolicySize
+  )
+{
+  EFI_STATUS  Status;
+
+  if ((PolicyInterface == NULL) ||
+      (GfxSiliconPolicy == NULL) || (PolicySize == NULL))
```
Comment (apop5) [inline Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c:66]: GfxSiliconPolicy is listed as an optional parameter. Is it optional?

File: Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+  {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (*PolicySize < GFX_POLICY_SIZE) {
+    return EFI_BUFFER_TOO_SMALL;
```
Comment (apop5) [inline Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c:72]: Policy size isn't updated, should it be updated with the required size?

File: Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c (hunk 9 above)
Comment (kuqin12) [reply Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c:66]:
```suggestion
      (PolicySize == NULL) || ((GfxSiliconPolicy == NULL) && (*PolicySize != 0)))
```
I think the check probably should be more like above?

File: Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c (hunk 10 above)
Comment (os-d) [reply Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c:72]: This was converted from an existing driver, with the new model I don't think this is useful to query the size, it is a static function and this module knows what size it should be (GFX_POLICY_SIZE).

There is no opportunity for this module to fix the size if the policy is incorrect.

@kuqin12 do you have any issue if I change the function to not have the expectation of querying the size?

**Review by apop5 — APPROVED** [review]

File: Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+  GFX_POLICY_DATA  GfxSiPol[GFX_PORT_MAX_CNT];
+  GFX_POLICY_DATA  GfxConfPol[GFX_PORT_MAX_CNT];
+
+  if ((PolicyInterface == NULL) ||
+      (ConfigBuffer == NULL))
+  {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  DEBUG ((DEBUG_ERROR, "%a Entry...\n", __FUNCTION__));
```
Comment (kuqin12) [inline Platforms/QemuQ35Pkg/ConfigKnobs/ConfigDataGfx/ConfigDataGfx.c:115]: nit: not error level?

**Review by kuqin12 — APPROVED** [review]

**Pull request merged**